### PR TITLE
feature: webhooks

### LINKED ISSUE #6778: If Kafka is turned on internal subs dont work
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME
~~~
API
~~~

##### CLOUDSTACK VERSION
~~~
4.17.0.1
~~~

##### SUMMARY
When kafka events are turned on, subscriptions internal to CloudStack can no longer receive events.
There appears to be code that is trying to use the events existing in the CloudStack code, but would stop working once kafka emitting is enabled.

This issue is to start a dialog about this functionality and/or directions.

A work around was created to resolve this issue, by creating an in-memory bus subscribe-able to anything internal to CloudStack and also emit events to any configured kafka endpoint simultaneously.  Thus allowing event processing internally to CloudStack without having to subscribe to external endpoint.

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Feature spec: https://cwiki.apache.org/confluence/display/CLOUDSTACK/Webhooks

Webhooks allow external services to be notified when certain events happen.

To implement and make Webhook work as an event bus, changes have been made to publish messages on multiple active message buses.
Fixes #6778 

Documentation PR: https://github.com/apache/cloudstack-documentation/pull/385

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### Screenshots (if appropriate):
Screenshot from 2024-02-20 13-54-49
Screenshot from 2024-02-20 13-55-19
Screenshot from 2024-02-20 13-56-31

### CODE REVIEW

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test keepEnv

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by GutoVeronezi** [comment]
@shwstppr do you have any description of the feature?

**Comment by shwstppr** [comment]
> @shwstppr do you have any description of the feature?

@GutoVeronezi yes, I've added a feature spec at https://cwiki.apache.org/confluence/display/CLOUDSTACK/Webhooks
This is still in WIP so some specific details may change over time.

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

File: framework/events/src/main/java/org/apache/cloudstack/framework/events/EventDistributorImpl.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    public void setEventBuses(List<EventBus> eventBuses) {
+        this.eventBuses = eventBuses;
+    }
+
+    @PostConstruct
+    public void init() {
+        if (logger.isTraceEnabled()) {
+            logger.trace(String.format("Found %d event buses : %s", eventBuses.size(),
+                    StringUtils.join(eventBuses.stream().map(x->x.getClass().getName()).toArray())));
+        }
```
Comment (DaanHoogland) [inline framework/events/src/main/java/org/apache/cloudstack/framework/events/EventDistributorImpl.java:45]:
```suggestion
        logger.trace("Found {} event buses : {}", eventBuses.size(),
                    () -> StringUtils.join(eventBuses.stream().map(x->x.getClass().getName()).toArray()));
```

might need testing, but it does seem simpler

File: framework/events/src/main/java/org/apache/cloudstack/framework/events/EventDistributorImpl.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        for (EventBus bus : eventBuses) {
+            try {
+                bus.publish(event);
+            } catch (EventBusException e) {
+                logger.warn(String.format("Failed to publish for bus %s of event [category: %s, type: %s]",
+                        bus.getName(), event.getEventCategory(), event.getEventType()));
+                if (logger.isTraceEnabled()) {
+                    logger.trace(event.getDescription());
+                }
```
Comment (DaanHoogland) [inline framework/events/src/main/java/org/apache/cloudstack/framework/events/EventDistributorImpl.java:67]: no parameter expansion so no guard needed
```suggestion
                logger.trace(event.getDescription());
```

Comment (DaanHoogland) [reply framework/events/src/main/java/org/apache/cloudstack/framework/events/EventDistributorImpl.java:67]: there is more of those. not commenting on any of those now.

**Review by DaanHoogland — COMMENTED** [review]
looks good so far @shwstppr . prios dictate that I do other work first, but will continue later.

**Comment by shwstppr** [comment]
@DaanHoogland thanks. Will address logging changes together with any remaining review comments

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan package

**Review by DaanHoogland — COMMENTED** [review]
CLGTM generally, the remarks about logging are basically only one remark; log4j v2 has different constraints. Apply my suggestions at will. One remark about a plural where I think singular is appropriate.

File: plugins/event-bus/webhook/src/main/java/org/apache/cloudstack/mom/webhook/WebhookApiServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+        String domainPath = null;
+        if (Account.Type.DOMAIN_ADMIN.equals(caller.getType())) {
+            Domain domain = domainDao.findById(caller.getDomainId());
+            domainPath = domain.getPath();
+        }
+        List<WebhookJoinVO> webhooks = webhookJoinDao.listByAccountOrDomain(caller.getId(), domainPath);
+        return webhooks.stream().map(WebhookJoinVO::getId).collect(Collectors.toList());
+    }
+
+    protected ManagementServerHostVO basicWebhookDeliveryApiCheck(Account caller, final Long id, final Long webhookId,
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply plugins/event-bus/webhook/src/main/java/org/apache/cloudstack/mom/webhook/WebhookApiServiceImpl.java:119]: This is protected to allow adding unit tests

File: framework/events/src/main/java/org/apache/cloudstack/framework/events/EventDistributorImpl.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+    List<EventBus> eventBuses;
+
+    public void setEventBuses(List<EventBus> eventBuses) {
+        this.eventBuses = eventBuses;
+    }
+
+    @PostConstruct
+    public void init() {
+        if (logger.isTraceEnabled()) {
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply framework/events/src/main/java/org/apache/cloudstack/framework/events/EventDistributorImpl.java]: @DaanHoogland yes

**Comment by shwstppr** [comment]
@blueorangutan package

File: plugins/event-bus/webhook/src/main/java/org/apache/cloudstack/mom/webhook/WebhookApiServiceImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply plugins/event-bus/webhook/src/main/java/org/apache/cloudstack/mom/webhook/WebhookApiServiceImpl.java:119]: would package private not be suitable for that?

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

File: engine/schema/src/main/resources/META-INF/db/schema-41900to42000.sql [hunk 5]
```diff
@@ -79,3 +78,46 @@ CREATE TABLE IF NOT EXISTS `cloud_usage`.`quota_email_configuration`(
...
+  `state` char(32) NOT NULL COMMENT 'state of the webhook - Enabled or Disabled',
+  `domain_id` bigint unsigned NOT NULL COMMENT 'id of the owner domain of the webhook',
+  `account_id` bigint unsigned NOT NULL COMMENT 'id of the owner account of the webhook',
+  `payload_url` varchar(255) COMMENT 'payload URL for the webhook',
+  `secret_key` varchar(255) COMMENT 'secret key for the webhook',
+  `ssl_verification` boolean COMMENT 'for https payload url, if true then strict ssl verification',
+  `scope` char(32) NOT NULL COMMENT 'scope for the webhook - Local, Domain, Global',
+  `created` datetime COMMENT 'date the webhook was created',
+  `removed` datetime COMMENT 'date removed if not null',
+  `generated_removed` BIGINT(15) GENERATED ALWAYS AS (IF(removed IS NULL, 0, TIMESTAMPDIFF(SECOND, from_unixtime(0), removed))),
```
Comment (shwstppr) [inline engine/schema/src/main/resources/META-INF/db/schema-41900to42000.sql:98]:
```suggestion
  `generated_removed` BIGINT(15) GENERATED ALWAYS AS (IF(removed IS NULL, 0, TIMESTAMPDIFF(SECOND, from_unixtime(0), removed))) PERSISTENT,
```

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
Retry

@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
Try again

@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan test rocky8 kvm-rocky8

**Comment by shwstppr** [comment]
@blueorangutan package

File: plugins/event-bus/webhook/src/main/java/org/apache/cloudstack/mom/webhook/WebhookApiServiceImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (shwstppr) [reply plugins/event-bus/webhook/src/main/java/org/apache/cloudstack/mom/webhook/WebhookApiServiceImpl.java:120]: that didn't work in my experience

**Comment by shwstppr** [comment]
@blueorangutan test rocky8 kvm-rocky8

**Comment by shwstppr** [comment]
Test failing with exceptions,
```
2024-05-28T13:43:38,730 WARN  [c.c.u.AccountManagerImpl] (API-Job-Executor-31:[ctx-26228571, job-39, ctx-aa487b81]) (logid:bf162f6e) Failed to cleanup account Account [{"accountName":"test-TestUserLogin-test_2FA_setup-CKUF5G","id":7,"uuid":"047e3b67-c492-48ae-b8ea-0b67358e1a64"}] due to com.cloud.utils.exception.CloudRuntimeException: DB Exception on: com.mysql.cj.jdbc.ClientPreparedStatement: UPDATE user_data SET user_data.removed='2024-05-28 13:43:38' WHERE user_data.account_id = 7
        at com.cloud.utils.db.GenericDaoBase.update(GenericDaoBase.java:871)
        at com.cloud.utils.db.GenericDaoBase.remove(GenericDaoBase.java:1993)
```
While https://github.com/apache/cloudstack/pull/9120 is merged, we need https://github.com/apache/cloudstack/pull/8893 for 4.19.1 DB changes to work in `main` cc @weizhouapache @DaanHoogland

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test rocky8 kvm-rocky8

**Comment by shwstppr** [comment]
@blueorangutan package

**Review by kiranchavala — APPROVED** [review]
LGTM , Tested manually and the webhook features works fine

Create a python webserver to

Followed this code 

https://gist.github.com/mdonkers/63e115cc0c79b4f6b8b3a6b797e485c7

or create a channel in https://smee.io/

Please find the test cases 


<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/kiranchavala/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/kiranchavala/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">


Test Cases | Result
-- | --
Test the   webhook delivery api calls (createWebhook,   listWebhook,updateWebhook,deleteWebhook,listWebhookDeliveries,deleteWebhookDelivery,executeWebhookDelivery) | Pass
Test the webhook at local, domain ,   global level, Project level | Pass
Test the https webhook with ssl   verfication | Pass
Test the webhook payload url uniqness at   account and domain level | Pass
Test the   global settings for webhook   (webhook.deliveries.limit,webhook.deliveries.cleanup.interval,webhook.delivery.thread.pool.size,   webhook.delivery.retries,webhook.delivery.timeout) | Pass
Test the webhook with a secret key | Pass



</body>

</html>

Sample post request 
```

10.0.35.105 - - [05/Jun/2024 07:23:05] "POST / HTTP/1.1" 200 -
INFO:root:POST request,
Path: /
Headers:
X-CS-Event-ID: 726a66f7-9aa4-495a-b2c4-b8217212d158
X-CS-Event: VM.START
User-Agent: CS-Hookshot/eb5a0506-0cfa-11ef-931c-1e00ae0003fc
Content-Length: 410
Content-Type: application/json; charset=UTF-8
Host: 10.0.32.251:8080
Connection: Keep-Alive
Accept-Encoding: gzip,deflate



Body:
{"eventDateTime":"2024-06-05 07:23:12 +0000","entityuuid":"c5be5c4e-b548-44f1-b585-297ed15954a9","description":"Successfully completed starting Vm. Vm Id: c5be5c4e-b548-44f1-b585-297ed15954a9","event":"VM.START","user":"eb5b1dce-0cfa-11ef-931c-1e00ae0003fc","account":"eb5a0506-0cfa-11ef-931c-1e00ae0003fc","entity":"VirtualMachine","status":"Completed","VirtualMachine":"c5be5c4e-b548-44f1-b585-297ed15954a9"}
```

File: plugins/event-bus/rabbitmq/src/main/java/org/apache/cloudstack/mom/rabbitmq/RabbitMQEventBus.java [hunk 6]
```diff
@@ -185,11 +185,14 @@ public static void setRetryInterval(Integer retryInterval) {
...
     @Override
     public UUID subscribe(EventTopic topic, EventSubscriber subscriber) throws EventBusException {
-
         if (subscriber == null || topic == null) {
             throw new EventBusException("Invalid EventSubscriber/EventTopic object passed.");
         }
 
+        if (logger.isDebugEnabled()) {
+            logger.debug("subscribing '{}' to events of type '{}' from '{}'", subscriber.toString(), topic.getEventType(), topic.getEventSource());
+        }
```
Comment (shwstppr) [inline plugins/event-bus/rabbitmq/src/main/java/org/apache/cloudstack/mom/rabbitmq/RabbitMQEventBus.java:194]:
```suggestion
        logger.debug("subscribing '{}' to events of type '{}' from '{}'", subscriber.toString(), topic.getEventType(), topic.getEventSource());
```

File: server/src/main/java/com/cloud/user/AccountManagerImpl.java [hunk 7]
```diff
@@ -425,6 +428,17 @@ public void setQuerySelectors(List<QuerySelector> querySelectors) {
...
     }
 
+    protected void deleteWebhooksForAccount(long accountId) {
+        try {
+            WebhookHelper webhookService = ComponentContext.getDelegateComponentOfType(WebhookHelper.class);
+            webhookService.deleteWebhooksForAccount(accountId);
+        } catch (NoSuchBeanDefinitionException ignored) {
+            if (logger.isDebugEnabled()) {
+                logger.debug("No WebhookHelper bean found");
+            }
```
Comment (shwstppr) [inline server/src/main/java/com/cloud/user/AccountManagerImpl.java:438]:
```suggestion
            logger.debug("No WebhookHelper bean found");
```

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by shwstppr** [comment]
@blueorangutan package

**Review by rohityadavcloud — APPROVED** [review]
LGTM, thanks for the feature efforts @shwstppr @kiranchavala

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

File: api/src/main/java/org/apache/cloudstack/api/command/admin/user/UpdateUserCmd.java [hunk 8]
```diff
@@ -66,7 +66,7 @@ public class UpdateUserCmd extends BaseCmd {
     @Parameter(name = ApiConstants.CURRENT_PASSWORD, type = CommandType.STRING, description = "Current password that was being used by the user. You must inform the current password when updating the password.", acceptedOnAdminPort = false)
     private String currentPassword;
 
-    @Parameter(name = ApiConstants.SECRET_KEY, type = CommandType.STRING, description = "The secret key for the user. Must be specified with userApiKey")
+    @Parameter(name = ApiConstants.USER_SECRET_KEY, type = CommandType.STRING, description = "The secret key for the user. Must be specified with userApiKey")
```
Comment (weizhouapache) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/user/UpdateUserCmd.java:69]: @shwstppr 
IMHO this should not be changed

File: api/src/main/java/org/apache/cloudstack/api/response/BucketResponse.java [hunk 9]
```diff
@@ -98,7 +98,7 @@ public class BucketResponse extends BaseResponseWithTagInformation implements Co
     @Param(description = "Bucket Access Key")
     private String accessKey;
 
-    @SerializedName(ApiConstants.SECRET_KEY)
+    @SerializedName(ApiConstants.USER_SECRET_KEY)
```
Comment (weizhouapache) [inline api/src/main/java/org/apache/cloudstack/api/response/BucketResponse.java:101]: same here, but not as important as previous one

File: api/src/main/java/org/apache/cloudstack/api/command/admin/user/UpdateUserCmd.java (hunk 8 above)
Comment (shwstppr) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/user/UpdateUserCmd.java:69]: @weizhouapache only the variable name is changed (adjusted to actual string value). API parameter is still the same - `usersecretkey`
https://github.com/apache/cloudstack/pull/8674/files#diff-72d5bf21c12ffd0a3c2d3a6033ec70ae5ba6c31338e915fe65df1b7a67827a9eL403

https://github.com/apache/cloudstack/pull/8674/files#diff-72d5bf21c12ffd0a3c2d3a6033ec70ae5ba6c31338e915fe65df1b7a67827a9eR491

File: api/src/main/java/org/apache/cloudstack/api/response/BucketResponse.java (hunk 9 above)
Comment (shwstppr) [reply api/src/main/java/org/apache/cloudstack/api/response/BucketResponse.java:101]: same as https://github.com/apache/cloudstack/pull/8674#discussion_r1636187875

File: api/src/main/java/org/apache/cloudstack/api/command/admin/user/UpdateUserCmd.java (hunk 8 above)
Comment (weizhouapache) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/user/UpdateUserCmd.java:69]: ok @shwstppr 
double-checked it, it is the same as before
was a bit confused.
sorry

**Comment by shwstppr** [comment]
@btzq I'm not sure if ACS currently publishes a separate event for router healthcheck fails but it does publish an event with type `ROUTER.HEALTH.CHECKS` which can be used in the webhook

**Comment by btzq** [comment]
@shwstppr under events, i only could find this:

Screenshot 2024-08-28 at 1 38 10 PM

But it seems to be under (Level = INFO) (Type = Health Check)

**Comment by btzq** [comment]
@shwstppr I think thats pretty good for us, we just need to have an alert when there is a failing health check. 

Right now, we have to login in ACS every few minutes and check the UI. We dont have RabbitMQ/Kafka installed.

**Comment by shwstppr** [comment]
@btzq that Type showing as `Health check` is a UI thing. If you check the API response it will have proper type,

```
    "listeventsresponse": {
        "count": 1,
        "event": [
            {
                "id": "36c995da-7b14-4083-9164-c486c7796b8c",
                "username": "system",
                "type": "ROUTER.HEALTH.CHECKS",
                "level": "INFO",
                "description": "Router r-9-VM / 9974ef30-b2fe-4c05-a2e3-57d3cfaf0b4b has failing checks: gateways_check.py",
                "account": "system",
                "domainid": "62a64961-4b15-11ef-953b-1e0054000142",
                "domain": "ROOT",
                "resourceid": "9974ef30-b2fe-4c05-a2e3-57d3cfaf0b4b",
                "resourcetype": "DomainRouter",
                "resourcename": "r-9-VM",
                "created": "2024-08-28T05:13:46+0000",
                "state": "Completed"
            }
        ]
    }
```

Maybe test creating a Global webhook and try there

**Comment by btzq** [comment]
I see, Thank you @shwstppr excited to test this!

**Pull request merged**